aptos-move/aptos-workspace-server/src/services/docker_common.rs

```diff
@@ -1,19 +1,69 @@
 // Copyright (c) Aptos Foundation
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::common::make_shared;
+use crate::common::{make_shared, ArcError};
 use anyhow::{anyhow, bail, Context, Result};
-use aptos_localnet::docker;
 use bollard::{
     container::{CreateContainerOptions, InspectContainerOptions, StartContainerOptions},
+    image::CreateImageOptions,
     network::CreateNetworkOptions,
     secret::ContainerInspectResponse,
     volume::CreateVolumeOptions,
+    Docker,
 };
+use futures::TryStreamExt;
 use std::{future::Future, sync::Arc};
 use tokio::sync::Mutex;
 use tokio_util::sync::CancellationToken;
 
+/// Creates a permanent docker network which does not need to be cleaned up.
+pub async fn create_docker_network_permanent(
+    shutdown: CancellationToken,
+    fut_docker: impl Future<Output = Result<Docker, ArcError>> + Clone + Send + 'static,
+    name: String,
+) -> Result<String, anyhow::Error> {
+    let handle = tokio::spawn(async move {
+        let docker = tokio::select! {
+            _ = shutdown.cancelled() => {
+                bail!("failed to create docker network: cancelled")
+            }
+            res = fut_docker => {
+                res.context("failed to create docker network")?
+            }
+        };
+
+        let res = docker
+            .create_network(CreateNetworkOptions {
+                name: name.clone(),
+                internal: false,
+                check_duplicate: true,
+                ..Default::default()
+            })
+            .await;
+
+        match res {
+            Ok(_response) => {
+                println!("Created docker network {}", name);
+
+                Ok(name)
+            },
+            Err(err) => match err {
+                bollard::errors::Error::DockerResponseServerError {
+                    status_code: 409, ..
+                } => {
+                    println!("Docker network {} already exists, not creating it", name);
+                    Ok(name)
+                },
+                err => Err(err.into()),
+            },
+        }
+    });
+
+    handle
+        .await
+        .map_err(|err| anyhow!("failed to join task handle: {}", err))?
```

Comment:
```suggestion
        .map_err(|err| anyhow!("failed to join task handle: {:#}", err))?
```
I think this is the best way to get the most context? Not 100% sure though.